messaging/api/src/main/java/org/eclipse/microprofile/reactive/messaging/MessagingProvider.java

```diff
@@ -0,0 +1,40 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package org.eclipse.microprofile.reactive.messaging;
+
+/**
+ * Marker interface to indicate which messaging provider should be used.
+ * <p/>
+ * If a container supports more than one messaging provider, then they may use this to differentiate between messaging
+ * providers by providing a subclass for each provider. Application streams can then use this subclass by passing it
+ * to the {@code provider} property of their {@link Incoming} or {@link Outgoing} annotations.
+ * <p/>
+ * Unless containers explicitly share libraries that provide the same messaging provider using the same sub class of
+ * this type, then messaging providers are assumed to be specific to each container. That is to say, if one container
+ * supports a messaging technology called Reactive Messaging, and provides a subclass of this called
+ * {@code org.example.container.ReactiveMessagingProvider}, and another container also supports the same technology,
+ * it won't necessarily use the same class, and from the containers point of view, the two different classes
+ * representing the same technology are for two different technologies.
+ *
+ * @see Incoming#provider()
+ * @see Outgoing#provider()
+ */
+public interface MessagingProvider {
```

Comment:
Currently, a messaging provider is selected based on the `provider` attribute on the `Incoming` and `Outgoing` annotations, hence, they can use different providers.

How a messaging provider is resolved, or even how it integrates with the implementation, is not specified, and is up to the implementation. In the implementation that I've done for this, there is a [`LightbendMessagingProvider`](https://github.com/lightbend/microprofile-reactive-messaging/blob/f807de040c7955c430dc48b18720bba25d9502e1/core/src/main/java/com/lightbend/microprofile/reactive/messaging/spi/LightbendMessagingProvider.java) interface (actually should probably be renamed to `AkkaMessagingProvider`), and messaging providers implement that. It returns a MicroProfile `MessagingProvider` class from it`s `providerFor` method. The Lightbend implementation then expects a messaging provider to provide an `ApplicationScoped` implementation of `LigtbendMessagingProvider`, and it then gathers all of the `LightbendMessagingProvider` implementations in the CDI container, and discovers which `MessagingProvider`'s it supports by invoking their `providerFor` method, see:

https://github.com/lightbend/microprofile-reactive-messaging/blob/f807de040c7955c430dc48b18720bba25d9502e1/core/src/main/java/com/lightbend/microprofile/reactive/messaging/impl/StreamManager.java#L28

So when it comes to actually using this, Lightbend's Kafka integration provides this implemnetation of `MessagingProvider`:

https://github.com/lightbend/microprofile-reactive-messaging/blob/f807de040c7955c430dc48b18720bba25d9502e1/kafka/src/main/java/com/lightbend/microprofile/reactive/messaging/kafka/Kafka.java

And that gets mapped to the actual implementation here:

https://github.com/lightbend/microprofile-reactive-messaging/blob/f807de040c7955c430dc48b18720bba25d9502e1/kafka/src/main/java/com/lightbend/microprofile/reactive/messaging/kafka/KafkaMessagingProvider.java#L67

But all of this is specific to Lightbend's implementation. Another implementation could well provide a mechanism based on `ServiceLoaders`. It's intentionally unspecified, specifying it would require creating an SPI for messaging providers, and I don't think we want to go down that path, at least not yet, perhaps a future version of this spec could do that.

Reply:
nice